python/tests/test_dbscan.py

```diff
@@ -0,0 +1,256 @@
+
+@pytest.mark.parametrize("feature_type", pyspark_supported_feature_types)
+@pytest.mark.parametrize("data_shape", [(1000, 20)], ids=idfn)
+@pytest.mark.parametrize("data_type", cuml_supported_data_types)
+@pytest.mark.parametrize("max_record_batch", [100, 10000])
+@pytest.mark.slow
+def test_dbscan_self(
+    gpu_number: int,
+    feature_type: str,
+    data_shape: Tuple[int, int],
+    data_type: np.dtype,
+    max_record_batch: int,
+) -> None:
+    """
+    The dataset of this test case comes from cuml:
+    https://github.com/rapidsai/cuml/blob/496f1f155676fb4b7d99aeb117cbb456ce628a4b/python/cuml/tests/test_kmeans.py#L39
+    """
+    from cuml.datasets import make_blobs
+
+    print("GPU_NUMBER", gpu_number)
+    n_rows = data_shape[0]
+    n_cols = data_shape[1]
+    n_clusters = 8
+    cluster_std = 1.0
+    tolerance = 0.001
+
+    eps = 5
+    min_samples = 5
+    metric = "euclidean"
+
+    X, _ = make_blobs(
+        n_rows, n_cols, n_clusters, cluster_std=cluster_std, random_state=0
+    )  # make_blobs creates a random dataset of isotropic gaussian blobs.
+
+    from cuml import DBSCAN as cuDBSCAN
+
+    cuml_dbscan = cuDBSCAN(
+        eps=eps, min_samples=min_samples, metric=metric, output_type="numpy", verbose=7
+    )
+
+    import cudf
+
+    gdf = cudf.DataFrame(X)
+    cuml_transformed = cuml_dbscan.fit_predict(gdf)
+
+    sample_to_cluster = dict()
+    cluster_dict: Dict[int, int] = dict()
+
+    np_df = X.get()
+    for rid, row in enumerate(np_df):
+        label = cuml_transformed[rid]
+
+        sample_to_cluster[tuple(row)] = label
+
+    conf = {"spark.sql.execution.arrow.maxRecordsPerBatch": str(max_record_batch)}
+    with CleanSparkSession(conf) as spark:
+        df, features_col, _ = create_pyspark_dataframe(
+            spark, feature_type, data_type, X, None
+        )
+
+        dbscan = DBSCAN(
+            num_workers=gpu_number,
+            eps=eps,
+            min_samples=min_samples,
+            metric=metric,
+            verbose=7,
+        ).setFeaturesCol(features_col)
+
+        dbscan_model = dbscan.fit(df)
+        dbscan_model.setPredictionCol("prediction")
+        transformed = dbscan_model.transform(df)
+
+        # Check cluster match
+        label_df = transformed.select("prediction")
+        feature_df = transformed.drop("prediction")
+
+        label_pdf = label_df.toPandas()
+        feature_pdf = feature_df.toPandas()
+
+        label_arr = label_pdf.to_numpy().squeeze()
+        feature_matrix = feature_pdf.to_numpy()
+
+        for rid, row in enumerate(feature_matrix):
+            if isinstance(row[0], DenseVector):
+                data = tuple(row[0].toArray())
+            elif isinstance(row[0], np.float32) or isinstance(row[0], np.float64):
+                data = tuple(row)
+            else:
+                data = tuple(row[0])
+
+            label_rapids = label_arr[rid]
+            label_cuml = sample_to_cluster[data]
+
+            if label_rapids in cluster_dict:
```

Comment:
Can we assert  "label_rapids == label_cuml" here?

Reply:
Another way is to adopt cuml tests that assert the labels of core points are the same, like here: https://github.com/rapidsai/cuml/blob/2c876b50be7422857806baedabbf66b498b56e21/python/cuml/tests/dask/test_dask_dbscan.py#L233

We may add a unique id column to spark dataframe to leverage cuml_dbscan.core_sample_indices_.

Reply:
Assert "label_rapids == label_cuml" does work, but I thought it's safer to do this cluster mapping in case there is discrepency between the exact cluster label even if the algorithm is working correctly. Should we swap to the simple assert equal?

Reply:
Nice way to address possible permutation of labels.